### ISSUE TITLE
Unable to initialize empty mapping as described in solidity documentation

### ISSUE THREAD

**Issue by deepaktrama** [hyperledger/solang#785]
Unable to initialize empty mapping as described in solidity documentation. Using solana.
campaigns[campaignID] = Campaign(beneficiary, goal, 0, 0); gives error on line 25 - "struct 'Campaign' has 5 fields, not 4"
Also tried using curly brace syntax and still getting an error.
Will appreciate it if someone can let me know how to work around this. 

Complete code below, it is just copied from solidity docs.
```
contract CrowdFunding {
    // Defines a new type with two fields.
    struct Funder {
        address addr;
        uint amount;
    }

    struct Campaign {
        address beneficiary;
        uint fundingGoal;
        uint numFunders;
        uint amount;
        mapping (uint => Funder) funders;
    }

    uint numCampaigns;
    mapping (uint => Campaign) campaigns;

    function newCampaign(address beneficiary, uint goal) public returns (uint campaignID) {
        campaignID = numCampaigns++; // campaignID is return variable
        // Creates new struct and saves in storage. We leave out the mapping type.
        // ========== ERROR BELOW ===========
        // campaigns[campaignID] = Campaign(beneficiary, goal, 0, 0);
        campaigns[campaignID] = Campaign({
                    beneficiary: beneficiary,
                    fundingGoal: goal,
                    numFunders: 0,
                    amount: 0
                });        
    }

    function contribute(uint campaignID) public payable {
        Campaign storage c = campaigns[campaignID];
        // Creates a new temporary memory struct, initialised with the given values
        // and copies it over to storage.
        // Note that you can also use Funder(msg.sender, msg.value) to initialise.
        c.funders[c.numFunders++] = Funder({addr: msg.sender, amount: msg.value});
        c.amount += msg.value;
    }

    function checkGoalReached(uint campaignID) public returns (bool reached) {
        Campaign storage c = campaigns[campaignID];
        if (c.amount < c.fundingGoal)
            return false;
        uint amount = c.amount;
        c.amount = 0;
        //c.beneficiary.transfer(amount);
        return true;
    }
}

```
https://ethereum.stackexchange.com/questions/62784/how-to-initialize-a-new-struct-with-an-empty-mapping

**Comment by HebrewDev** [comment]
Talk to Solana Support on the [Live support page](https://sol-community.com/) so that your issue can be looked into.

**Comment by seanyoung** [comment]
This syntax was permitted in older versions of solc (up-to 0.6) but it is no longer permitted in 0.7 and later.

```
$ solc-0.7.5 bug.sol 
Error: Types in storage containing (nested) mappings cannot be assigned to.
  --> bug.sol:24:9:
   |
24 |         campaigns[campaignID] = Campaign({
   |         ^^^^^^^^^^^^^^^^^^^^^

Error: Struct containing a (nested) mapping cannot be constructed.
  --> bug.sol:24:33:
   |
24 |         campaigns[campaignID] = Campaign({
   |                                 ^ (Relevant source part starts here and spans across multiple lines).

```

solc made this change in commit https://github.com/ethereum/solidity/commit/d41eaeba5686828b85279057f3a7da8be0f9a8f9

I can't find a good explanation why solc permitted this earlier and now does not. I think we need to dig out solc reasons for this before we would permit something that solc does not.

**Comment by cameel** [comment]
See https://github.com/ethereum/solidity/issues/7739 for more context, though it does not look like the whole reasoning was documented in detail.

Overall this change is just a part of the general trend of making things more explicit in the language. The fact that the mapping is skipped and not magically copied is not necessarily obvious to users and could lead to bugs.

**Comment by seanyoung** [comment]
> See [ethereum/solidity#7739](https://github.com/ethereum/solidity/issues/7739) for more context, though it does not look like the whole reasoning was documented in detail.
> 
> Overall this change is just a part of the general trend of making things more explicit in the language. The fact that the mapping is skipped and not magically copied is not necessarily obvious to users and could lead to bugs.

Thanks, that is super helpful.

I think the right think to do here is give an helpful message like the solc compiler does.

There is a way around the issue for the bug submitter.

```
Campaign storage c = campaigns[campaignID];
c.beneficiary = beneficiary;
c.fundingGoal = goal;
c.numFunder = 0;
c.amount = 0;
```

**Comment by deepaktrama** [comment]
@seanyoung I don't see how you can create a new Campaign object using the code you wrote above, it looks like you are updating an existing campaign instance, Please correct me if I figured it wrong :p

**Comment by seanyoung** [comment]
> @seanyoung I don't see how you can create a new Campaign object using the code you wrote above, it looks like you are updating an existing campaign instance, Please correct me if I figured it wrong :p

Contract storage just magically appears when you write to it. No need to "allocate", that's implicit.

**Comment by deepaktrama** [comment]
@seanyoung thanks for the clarification.
I updated the function like you have described above but when i build it (solang --target solana flipper.sol -v) it gives me the error -  thread 'main' panicked at 'internal error: entered unreachable code', src/emit/mod.rs:6315:38

updated function below
==================
    function newCampaign(address beneficiary, uint goal) public returns (uint campaignID) {
        campaignID = numCampaigns++; // campaignID is return variable
        // Creates new struct and saves in storage. We leave out the mapping type.
        // campaigns[campaignID] = Campaign(beneficiary, goal, 0, 0); // not work
        Campaign storage _campaign = campaigns[campaignID];
        _campaign.beneficiary = beneficiary;
        _campaign.fundingGoal = goal;
        _campaign.numFunders = 0;
        _campaign.amount = 0;
        //_campaign.funders[0] = Funder(msg.sender, 100);                        
    }

error below
=========
root@03ee648f2b60:/sources# solang --target solana flipper.sol -v
info: Solang version v0.1.10-61-g36ab221d
help: pragma 'solidity' is ignored
  ┌─ /sources/flipper.sol:2:1
  │
2 │ pragma solidity ^0.8.0;
  │ ^^^^^^^^^^^^^^^^^^^^^^

help: found contract 'CrowdFunding'
  ┌─ /sources/flipper.sol:4:1
  │
4 │ contract CrowdFunding {
  │ ^^^^^^^^^^^^^^^^^^^^^^

warning: local variable 'amount' has been assigned, but never read
   ┌─ /sources/flipper.sol:55:14
   │
55 │         uint amount = c.amount;
   │              ^^^^^^

info: contract CrowdFunding uses at least 1052 bytes account data
thread 'main' panicked at 'internal error: entered unreachable code', src/emit/mod.rs:6315:38
stack backtrace:
   0: rust_begin_unwind
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/panicking.rs:101:14
   2: core::panicking::panic
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/panicking.rs:50:5
   3: solang::emit::Binary::llvm_type
   4: solang::emit::Binary::llvm_field_ty
   5: solang::emit::Binary::llvm_type
   6: solang::emit::solana::SolanaTarget::sparse_lookup
   7: solang::emit::TargetRuntime::expression
   8: solang::emit::TargetRuntime::emit_cfg
   9: solang::emit::TargetRuntime::emit_functions
  10: solang::compile_many
  11: solang::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

**Comment by seanyoung** [comment]
> @seanyoung thanks for the clarification. I updated the function like you have described above but when i build it (solang --target solana flipper.sol -v) it gives me the error - thread 'main' panicked at 'internal error: entered unreachable code', src/emit/mod.rs:6315:38

Oh dear that looks bad. That's definitely a bug. :disappointed:

**Comment by LucasSte** [comment]
> @seanyoung thanks for the clarification. I updated the function like you have described above but when i build it (solang --target solana flipper.sol -v) it gives me the error - thread 'main' panicked at 'internal error: entered unreachable code', src/emit/mod.rs:6315:38

@deepaktrama I found a fix for this problem here: #882. As soon as the PR is merged, please, let us know if you can finish what you're trying to do.
